tensorflow/python/framework/tensor_shape.py

```diff
@@ -942,19 +941,18 @@ def merge_with(self, other):
     other = as_shape(other)
     if self._dims is None:
       return other
+    if other.dims is None:
+      return self
     else:
       try:
         self.assert_same_rank(other)
-        new_dims = []
-        for i, dim in enumerate(self._dims):
-          new_dims.append(dim.merge_with(other[i]))
+        new_dims = [dim.merge_with(other_dim)
+                    for dim, other_dim in zip(self._dims, other.dims)]
         return TensorShape(new_dims)
       except ValueError:
         raise ValueError("Shapes %s and %s are not compatible" % (self, other))
 
   def __add__(self, other):
-    if not isinstance(other, TensorShape):
-      other = TensorShape(other)
```

Comment:
Let's not remove this unless we're absolutely sure it's a safe change.

Reply:
I think this is save to change since the first step in `self.concatenate` is to convert `other` into a `TensorShape` using:
https://github.com/tensorflow/tensorflow/blob/fd3c782277bee45ee496d6fb5ca440103e2101cf/tensorflow/python/framework/tensor_shape.py#L982
which handles exactly that case:
https://github.com/tensorflow/tensorflow/blob/fd3c782277bee45ee496d6fb5ca440103e2101cf/tensorflow/python/framework/tensor_shape.py#L1230-L1235

Reply:
Nice! lgtm.